pkg/pillar/cmd/zedagent/zedagent.go

```diff
@@ -291,323 +435,764 @@ func Run(ps *pubsub.PubSub, loggerArg *logrus.Logger, logArg *base.LogObject) in
-	getconfigCtx := getconfigContext{
+	// Initialize context used to get and parse device configuration.
+	getconfigCtx := &getconfigContext{
 		localServerMap: &localServerMap{},
 		// default value of currentMetricInterval
 		currentMetricInterval: zedagentCtx.globalConfig.GlobalValueInt(types.MetricInterval),
 		// edge-view configure
 		configEdgeview: &types.EdgeviewConfig{},
 	}
-	cipherCtx := cipherContext{}
-	attestCtx := attestContext{}
 
-	// Pick up (mostly static) AssignableAdapters before we report
-	// any device info
-	aa := types.AssignableAdapters{}
-	zedagentCtx.assignableAdapters = &aa
+	cipherCtx := &cipherContext{}
+	attestCtx := &attestContext{}
+	dnsCtx := &DNSContext{}
+	zedagentCtx.dnsCtx = dnsCtx
 
-	// Cross link
-	getconfigCtx.zedagentCtx = &zedagentCtx
-	zedagentCtx.getconfigCtx = &getconfigCtx
+	// Cross links between contexts.
+	getconfigCtx.zedagentCtx = zedagentCtx
+	zedagentCtx.getconfigCtx = getconfigCtx
 
-	cipherCtx.zedagentCtx = &zedagentCtx
-	zedagentCtx.cipherCtx = &cipherCtx
+	cipherCtx.zedagentCtx = zedagentCtx
+	zedagentCtx.cipherCtx = cipherCtx
 
-	attestCtx.zedagentCtx = &zedagentCtx
-	zedagentCtx.attestCtx = &attestCtx
+	attestCtx.zedagentCtx = zedagentCtx
+	zedagentCtx.attestCtx = attestCtx
+	return zedagentCtx
+}
 
-	pubZedAgentStatus, err := ps.NewPublication(pubsub.PublicationOptions{
-		AgentName: agentName,
-		TopicType: types.ZedAgentStatus{},
-	})
-	if err != nil {
-		log.Fatal(err)
+func initializeDirs() {
+	// create persistent holder directory
+	if _, err := os.Stat(types.PersistDir); err != nil {
+		log.Tracef("Create %s", types.PersistDir)
+		if err := os.MkdirAll(types.PersistDir, 0700); err != nil {
+			log.Fatal(err)
+		}
 	}
-	pubZedAgentStatus.ClearRestarted()
-	getconfigCtx.pubZedAgentStatus = pubZedAgentStatus
-
-	// apply saved radio config ASAP
-	initializeRadioConfig(&getconfigCtx)
-
-	// Wait until we have been onboarded aka know our own UUID
-	subOnboardStatus, err := ps.NewSubscription(pubsub.SubscriptionOptions{
-		AgentName:     "zedclient",
-		MyAgentName:   agentName,
-		TopicImpl:     types.OnboardingStatus{},
-		Activate:      true,
-		Persistent:    true,
-		Ctx:           &zedagentCtx,
-		CreateHandler: handleOnboardStatusCreate,
-		ModifyHandler: handleOnboardStatusModify,
-		WarningTime:   warningTime,
-		ErrorTime:     errorTime,
-	})
-	if err != nil {
-		log.Fatal(err)
+	if _, err := os.Stat(types.CertificateDirname); err != nil {
+		log.Tracef("Create %s", types.CertificateDirname)
+		if err := os.MkdirAll(types.CertificateDirname, 0700); err != nil {
+			log.Fatal(err)
+		}
+	}
+	if _, err := os.Stat(checkpointDirname); err != nil {
+		log.Tracef("Create %s", checkpointDirname)
+		if err := os.MkdirAll(checkpointDirname, 0700); err != nil {
+			log.Fatal(err)
+		}
 	}
-	// Wait for Onboarding to be done by client
+}
+
+func waitUntilOnboarded(zedagentCtx *zedagentContext, stillRunning *time.Ticker) {
+	zedagentCtx.subOnboardStatus.Activate()
```

Comment:
It is quite strange that yetus ignore this, but should we check error from `Activate()` functions? At least we can show error in logs.

Reply:
I checked and all errors that Activate may return are logged by the function.
What we could do additionally here is to trigger log.Fatal, but I don't know if it is a good idea.